domain/modelagent/state/modelstate.go

```diff
@@ -1051,6 +1099,154 @@ UPDATE agent_version SET stream_id = $agentVersionStream.stream_id
 	})
 }
 
+// SetModelTargetAgentVersion is responsible for setting the current target
+// agent version of the model. This function expects a precondition version to
+// be supplied. The model's target agent version at the time the operation is
+// applied must match the preCondition version or else an error is returned.
+func (st *State) SetModelTargetAgentVersion(
+	ctx context.Context,
+	preCondition semversion.Number,
+	toVersion semversion.Number,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	checkAgentVersionStmt, err := st.Prepare(`
+SELECT &agentVersionTarget.*
+FROM   agent_version
+`,
+		agentVersionTarget{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	toVersionInput := setAgentVersionTarget{TargetVersion: toVersion.String()}
+	setAgentVersionStmt, err := st.Prepare(`
+UPDATE agent_version
+SET    target_version = $setAgentVersionTarget.target_version
+`,
+		toVersionInput,
+	)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	preConditionVersionStr := preCondition.String()
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		currentAgentVersion := agentVersionTarget{}
+		err := tx.Query(ctx, checkAgentVersionStmt).Get(&currentAgentVersion)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return errors.New(
+				"checking current target agent version for model, no agent version has been set previously",
+			)
+		} else if err != nil {
+			return errors.Errorf(
+				"checking current target agent version for model: %w", err,
+			)
+		}
+
+		if currentAgentVersion.TargetVersion != preConditionVersionStr {
+			return errors.Errorf(
+				"unable to set agent version for model. The agent version has changed to %q",
```

Comment:
I guess we are doing the pre condition check as a separate query rather than a a single query using "update where" and checking the affected rows is so that the altered version can be reported?

Reply:
Also, I'm wondering if this should be a typed error. I can imagine the case where 2 users run upgrade-model at the same time; the typed error would be used to inform the generation of an error message specific to this mode of failure, instead of effectively recycling the error generated here.